File: coding-test-success/ch02/06_실패율.java
import java.util.HashMap;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] levels = new int[N + 2];
        
        // 스테이지별 사람 수 구하기
        for (int i = 0; i < stages.length; i++) {
            levels[stages[i]] += 1;
        }
        
        HashMap<Integer, Double> fails = new HashMap<>();
        double total = stages.length;
        
        // 레벨을 순회하며 실패율 계산
        for (int i = 1; i <= N; i++) {
            if(levels[i] == 0) {
                fails.put(i, 0.);
            } else {
                fails.put(i, levels[i] / total);
                total -= levels[i];
            }
        }
        // 실패율이 높은 스테이지부터 내림차순 정렬
        return fails.entrySet().stream().sorted((o1,o2) -> Double.compare(o2.getValue(),o1.getValue()))
            .mapToInt(HashMap.Entry::getKey).toArray();
    }
}

File: coding-test-success/ch02/06_실패율.py
def solution(N, stages):
    giveup = {}
    total = len(stages)
    
    levels = [0] * (N + 2)  # N+1까지 수용하기 위해 N+2로 설정
    
    # 각 레벨별 인원 계산
    for stage in stages:
        if stage <= N:
            levels[stage] += 1
    
    # 실패율 계산
    for i in range(1, N + 1):
        if total > 0:
            fail_rate = levels[i] / total
            giveup[i] = fail_rate
            total -= levels[i]
        else:
            giveup[i] = 0  # 스테이지에 도달한 유저가 없는 경우 실패율을 0으로 설정
    
    # 실패율 내림차순 및 스테이지 번호 오름차순 정렬
    ans = sorted(giveup, key=lambda x: (-giveup[x], x))
    
    return ans
